src/lib.rs

```diff
@@ -129,6 +129,106 @@ impl<'a> From<PrivatePkcs8KeyDer<'a>> for PrivateKeyDer<'a> {
+        //   }
+        // PKCS#5 (https://www.rfc-editor.org/rfc/rfc8018) describes the AlgorithmIdentifier
+        // as a SEQUENCE.
+        //
+        // Therefore, we consider the outer SEQUENCE, a Version of 0, and the start of
+        // an AlgorithmIdentifier to be enough to identify a PKCS#8 key. If it were PKCS#1,
+        // the version would not be followed by a SEQUENCE. If it were SEC1, the version would
+        // not have been 0.
+        let pkcs8_hint = [TAG_INTEGER, 0x01, 0x00, TAG_SEQUENCE];
+
+        // PKCS#1 (https://www.rfc-editor.org/rfc/rfc8017) describes the RSAPrivateKey structure
+        // as:
+        //  RSAPrivateKey ::= SEQUENCE {
+        //              version           Version,
+        //              modulus           INTEGER,  -- n
+        //              publicExponent    INTEGER,  -- e
+        //              privateExponent   INTEGER,  -- d
+        //              prime1            INTEGER,  -- p
+        //              prime2            INTEGER,  -- q
+        //              exponent1         INTEGER,  -- d mod (p-1)
+        //              exponent2         INTEGER,  -- d mod (q-1)
+        //              coefficient       INTEGER,  -- (inverse of q) mod p
+        //              otherPrimeInfos   OtherPrimeInfos OPTIONAL
+        //          }
+        //
+        // Therefore, we consider the outer SEQUENCE and a Version of 0 to be enough to identify
+        // a PKCS#1 key. If it were PKCS#8, the version would be followed by a SEQUENCE. If it
+        // were SEC1, the VERSION would have been 1.
+        let pkcs1_hint = [TAG_INTEGER, 0x01, 0x00];
+
+        // SEC1 (https://www.rfc-editor.org/rfc/rfc5915) describes the ECPrivateKey structure as:
+        //   ECPrivateKey ::= SEQUENCE {
+        //      version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+        //      privateKey     OCTET STRING,
+        //      parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
+        //      publicKey  [1] BIT STRING OPTIONAL
+        //   }
+        //
+        // Therefore, we consider the outer SEQUENCE and an INTEGER of 1 to be enough to
+        // identify a SEC1 key. If it were PKCS#8 or PKCS#1, the version would have been 0.
+        let sec1_hint = [TAG_INTEGER, 0x01, 0x01];
+
+        Ok(if key_bytes.starts_with(&pkcs8_hint) {
+            Self::Pkcs8(key.into())
+        } else if key_bytes.starts_with(&pkcs1_hint) {
+            Self::Pkcs1(key.into())
+        } else if key_bytes.starts_with(&sec1_hint) {
+            Self::Sec1(key.into())
+        } else {
+            return Err(INVALID_KEY_DER_ERR);
+        })
+    }
+}
+
+static INVALID_KEY_DER_ERR: &str = "unknown or invalid key format";
+
+#[cfg(feature = "alloc")]
+impl<'a> TryFrom<Vec<u8>> for PrivateKeyDer<'a> {
+    type Error = &'static str;
+
+    fn try_from(key: Vec<u8>) -> Result<Self, Self::Error> {
+        Ok(match PrivateKeyDer::try_from(&key[..])? {
```

Comment:
Is it on purpose that you're using `PrivateKeyDer` here instead of `Self`? It seems a little weird when `Self` is `PrivateKeyDer` to use both (but maybe this is one of those cases where the lifetime thing interferes with using `Self`).

Reply:
full disclosure: I took this part from the original PR without much thought :-)

>  maybe this is one of those cases where the lifetime thing interferes with using Self

I believe that's the case. Making a quick swap I get a `key` does not live long enough error.